[herd]: CMP (shifted register) barrel shifters

lukeg101 opened the pull request:
Consider the following AArch64 instruction:
```
CMP WZR, w1, asr #1;
CMP XZR, x1, asr #1;
```
which arithmetic shifts right the value in w1/x1 by 1 before
comparing with xzr/wzr. This is equivalent to:
```
SUBS XZR, WZR, w1, asr #1;
SUBS XZR, XZR, x1, asr #1;
```
This patch adds:
- barrel shifters to the CMP command
- tests for CMP with barrel shifters
- tests for the equivalent SUBS instructions

tested with:
- pre-commit tests - all pass
- make test - all pass
- make version - all pass/no warnings

maranget commented:
Hi @lukeg101,
A simple, yet useful PR. Would not instruction printing (`do_print_instruction` in `lib/AArch64Base.ml`) and litmus output (`litmus/AArch64Compile.ml`) also use the `cmp` form?

lukeg101 commented:
> Hi @lukeg101,
> A simple, yet useful PR. Would not instruction printing (`do_print_instruction` in `lib/AArch64Base.ml`) and litmus output (`litmus/AArch64Compile.ml`) also use the `cmp` form?

Good spot thanks for that, I have updated both files with appropriate pretty printing for `CMP`. I have also rebased to newer code - all tests pass

relokin commented on herd/tests/instructions/AArch64/A164.litmus:
```diff
@@ -0,0 +1,13 @@
+  MOV X0, #1;
+  foo: NOP;
+
+forall (0:x1=1 /\ 0:XZR=0 /\ 0:X0=0)
```
s/0:x1=1/0:X1=1/

lukeg101 replied:
done

relokin commented on herd/tests/instructions/AArch64/A165.litmus:
```diff
@@ -0,0 +1,14 @@
+ { 0:X1=1; }
+
+P0;
+  CMP WZR, w1, asr #1;
```
s/w1/W1

lukeg101 replied:
done

relokin commented on herd/tests/instructions/AArch64/A165.litmus:
```diff
@@ -0,0 +1,14 @@
+  MOV X0, #1;
+  foo: NOP;
+
+forall (0:x1=1 /\ 0:XZR=0 /\ 0:X0=0)
```
s/0:x1=1/0:X1=1/

lukeg101 replied:
done

relokin commented on herd/tests/instructions/AArch64/A166.litmus:
```diff
@@ -0,0 +1,13 @@
+ { 0:X1=1; }
+
+P0;
+  SUBS XZR, XZR, x1, asr #1;
```
s/x1/X1/

lukeg101 replied:
done

relokin commented on herd/tests/instructions/AArch64/A166.litmus:
```diff
@@ -0,0 +1,13 @@
+  MOV X0, #1;
+  foo: NOP;
+
+forall (0:x1=1 /\ 0:XZR=0 /\ 0:X0=0)
```
s/x1/X1/

lukeg101 replied:
done

relokin commented on herd/tests/instructions/AArch64/A167.litmus:
```diff
@@ -0,0 +1,13 @@
+ { 0:X1=1; }
+
+P0;
+  SUBS WZR, WZR, w1, asr #1;
```
s/w1/W1/

lukeg101 replied:
done

relokin commented on herd/tests/instructions/AArch64/A167.litmus:
```diff
@@ -0,0 +1,13 @@
+  MOV X0, #1;
+  foo: NOP;
+
+forall (0:x1=1 /\ 0:XZR=0 /\ 0:X0=0)
```
s/0:x1=1/0:X1=1/

lukeg101 replied:
done

relokin reviewed (commented):
Thanks Luke!

Looks good to me, I only have a couple of minor styling nits, feel free to ignore but I think it would be good if we could minimize the inconsistencies on the register naming conventions. Comments on the code.

For the commit message you might want to change

`SUBS XZR, WZR, w1, asr #1;`

to 

`SUBS WZR, WZR, W1, ASR #1;`

lukeg101 commented:
Hi Nikos, 
Thanks for taking a look at this, I have made these changes and updated the hashes in the `expected` files.
Please let me know if there is anything else.
Thanks
Luke

maranget commented on herd/tests/instructions/AArch64/A164.litmus:
```diff
@@ -0,0 +1,13 @@
+ { 0:X1=1; }
+
+P0;
+  CMP XZR, x1, ASR #1;
```
Small x1 forgotten:
```suggestion
  CMP XZR, X1, ASR #1;
```

lukeg101 replied:
good spot, fixed

maranget commented on herd/tests/instructions/AArch64/A164.litmus:
```diff
@@ -0,0 +1,13 @@
+AArch64 A164
+(* CMP shifted register is equivalent to SUBS WZR *)
+(* tests xreg *)
+
+ { 0:X1=1; }
+
+P0;
+  CMP XZR, X1, ASR #1;
+  B.EQ foo;
+  MOV X0, #1;
+  foo: NOP;
+
+forall (0:X1=1 /\ 0:XZR=0 /\ 0:X0=0)
```
This test is slightly incorrect as regards printing rules, litmus and herd print registers according to their types. Here it does not harm but it gives me the opportunity to detail the connections  between types and printing.

Implicitly this type is `int` (32bits). Here all registers are used in X form, and hence are meant to contain 64bit integers. This is no big deal because the values in registers are small positive integer. This would be a problem if, for instance 0:X1 would contain 2^32, then it will be printed as zero and not as 4294967296.
Finally to avoid such (infrequent) surprises, I strive to use W registers or to declare appropriate types for printed registers (here `int64_t 0:X1; int64_t 0:X0;`, It does not seem very useful for XZR).

maranget reviewed (approved):
I am ready to merge this PR, please consider the comment on types.

maranget commented:
Hi @lukeg101 , I have merged this PR by hand. Hene the explicit closing.

Thanks
